File: test-suite-kotlin/src/test/kotlin/io/micronaut/kafka/docs/consumer/topics/FavoriteEventListener.kt
package io.micronaut.kafka.docs.consumer.topics

import io.micronaut.configuration.kafka.annotation.KafkaKey
import io.micronaut.configuration.kafka.annotation.KafkaListener
import io.micronaut.configuration.kafka.annotation.Topic
import io.micronaut.context.annotation.Requires

@Requires(property = "spec.name", value = "FavoriteEventListenerTest")
@KafkaListener
class FavoriteEventListener {

    // tag::commonSupertype[]
    @Topic("favorites-events")
    fun receive(@KafkaKey customerId: String, event: FavoriteEvent) {
        when (event) {
            is FavoriteSaved -> {
                // process the save event for this customer
            }
            is FavoriteDeleted -> {
                // process the delete event for this customer
            }
        }
    }
    // end::commonSupertype[]
}

sealed interface FavoriteEvent

class FavoriteSaved : FavoriteEvent

class FavoriteDeleted : FavoriteEvent

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/topics/FavoriteEventListener.py
from abc import ABC
from typing import Annotated

from micronaut.configuration.kafka.annotation import KafkaKey, KafkaListener, Topic
from micronaut.context.annotation import Requires


class FavoriteEvent(ABC):
    pass


class FavoriteSaved(FavoriteEvent):
    pass


class FavoriteDeleted(FavoriteEvent):
    pass


@Requires(property="spec.name", value="FavoriteEventListenerTest")
@KafkaListener
class FavoriteEventListener:

    # tag::commonSupertype[]
    @Topic("favorites-events")
    def receive(self, customer_id: Annotated[str, KafkaKey], event: FavoriteEvent) -> None:
        if isinstance(event, FavoriteSaved):
            # process the save event for this customer
            pass
        elif isinstance(event, FavoriteDeleted):
            # process the delete event for this customer
            pass
    # end::commonSupertype[]
